### PR TITLE
Docstrings for arraymodule

### PR DESCRIPTION
**Pull request by e-marshall**
[Description of PR]
This is a draft PR that adds docstrings and a few code examples to the `Array` module. 
It's not complete, but I probably won't have more time to work on this today. I'd also love to check in and make sure things look okay. Any feedback or suggestions are very welcome! 

TODO:
* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [ ] Changes documented in docs/release.rst
* [ ] GitHub Actions have all passed
* [ ] Test coverage is 100% (Codecov passes)

### CODE REVIEW

File: src/zarr/core/array.py [hunk 1]
```diff
@@ -161,6 +161,77 @@ async def get_array_metadata(
...
+    metadata : ArrayMetadata
+        The metadata of the array.
+    store_path : StorePath
+        The path to the Zarr store.
+    codec_pipeline : CodecPipeline
+        The codec pipeline used for encoding and decoding chunks.
+    order : {'C', 'F'}
+        The order of the array data in memory.
+
+    Methods
```
Comment (TomAugspurger) [inline src/zarr/core/array.py:189]: I think this Methods section can be removed. Usually it's only used if you want to highlight a few methods on a class with many methods. I suspect that we want everything and use sphinx's autodoc to link everything together.

File: src/zarr/core/array.py [hunk 2]
```diff
@@ -161,6 +161,77 @@ async def get_array_metadata(
...
+    open(store, ...)
+        Open an existing AsyncArray from the given store.
+    resize(new_shape, ...)
+        Resize the AsyncArray to the specified shape.
+    update_attributes(new_attributes, ...)
+        Update the attributes of the AsyncArray.
+    info()
+        Print information about the AsyncArray.
+
+    Properties
```
Comment (TomAugspurger) [inline src/zarr/core/array.py:202]: https://numpydoc.readthedocs.io/en/latest/format.html#class-docstring mentions that the docstrings for properties can go on the property, and the class just needs their name.

File: src/zarr/core/array.py [hunk 3]
```diff
@@ -218,6 +289,69 @@ async def create(
...
+        Returns
+        -------
+        AsyncArray
+            The created asynchronous array instance.
+
+        Examples
+        --------
+        >>> import zarr
+        >>> store = zarr.store.MemoryStore(mode='w')
+        >>> async_arr = await zarr.core.array.AsyncArray.create(
```
Comment (TomAugspurger) [inline src/zarr/core/array.py:346]: Subsequent lines start with a `...` and then the indentation.:

```suggestion
>>> async_arr = await zarr.core.array.AsyncArray.create(
...     store=store,
...     shape=(100,100),
...     chunks=(10,10),
...     dtype='i4',
...     fill_value=0,
... )
```

File: src/zarr/core/array.py [hunk 4]
```diff
@@ -218,6 +289,69 @@ async def create(
...
+        >>>     store=store,
+        >>>     shape=(100,100),
+        >>>     chunks=(10,10),
+        >>>     dtype='i4',
+        >>>     fill_value=0)
+        <AsyncArray memory://140349042942400 shape=(100, 100) dtype=int32>
```
Comment (TomAugspurger) [inline src/zarr/core/array.py:352]: Usually a good idea to remove things that are platform / session dependent, like the memory address of the array:

```suggestion
        <AsyncArray memory://... shape=(100, 100) dtype=int32>
```

You can include a `# doctest: +ELLIPSIS` at the end of the line and it should be runnable through doctest, but I don't think we're running doctests on the project right now.

**Review by TomAugspurger — COMMENTED** [review]
Looks great, thanks! Just a couple small notes.

**Comment by e-marshall** [comment]
Thanks so much @TomAugspurger !

File: src/zarr/core/array.py (hunk 1 above)
Comment (e-marshall) [reply src/zarr/core/array.py:189]: okay, thanks!

File: src/zarr/core/array.py (hunk 2 above)
Comment (e-marshall) [reply src/zarr/core/array.py:202]: thanks! I'd missed this

File: src/zarr/core/array.py (hunk 4 above)
Comment (e-marshall) [reply src/zarr/core/array.py:352]: okay, thanks! So not running doctests for now, but fine to have in there for the future? I added it in but can remove if needed

**Comment by e-marshall** [comment]
https://github.com/zarr-developers/zarr-python/pull/2276/commits/461b575745b492ef662d5e42d6c62cdee5aa89d8 incorporates @TomAugspurger 's comments above. i got a mypy error on precommit that line 610 in `array.py` is unreachable. I'm a bit confused on this because it a) wasn't raised before, b) I think it should still be possible for `self.metadata.attributes` to be `None`, wondering if anyone has any clues on this?
```
600        @property
601         def attrs(self) -> dict[str, JSON]:
602        """Returns the attributes of the array.

604        Returns
605        -------
606        dict
607            Attributes of the array
608        """
609        if self.metadata.attributes is None:
610            return {}
611        return self.metadata.attributes
```

File: src/zarr/core/array.py [hunk 5]
```diff
@@ -218,6 +261,69 @@ async def create(
...
+
+        Returns
+        -------
+        AsyncArray
+            The created asynchronous array instance.
+
+        Examples
+        --------
+        >>> import zarr
+        >>> store = zarr.store.MemoryStore(mode='w')
```
Comment (MSanKeys963) [inline src/zarr/core/array.py:317]:
```suggestion
        >>> store = zarr.storage.MemoryStore(mode='w')
```
I think it should be `storage`. 
Check: https://github.com/zarr-developers/zarr-python/blob/v3/src/zarr/storage/memory.py

Comment (e-marshall) [reply src/zarr/core/array.py:317]: oops, thanks!

**Comment by jhamman** [comment]
@e-marshall - this is still marked as a "work in progress"  -- is it ready for a review?

**Comment by e-marshall** [comment]
sorry, got sidetrackd last week! Thanks for the nudge @jhamman , this is ready for review

**Review by dstansby — CHANGES_REQUESTED** [review]
I left one comment for a section that can be removed. This also needs a rebase to fix the merge conflict. Otherwise, at a glance it looks good to me 👍

File: src/zarr/core/array.py [hunk 6]
```diff
@@ -162,6 +162,49 @@ async def get_array_metadata(
...
+    metadata : ArrayMetadata
+        The metadata of the array.
+    store_path : StorePath
+        The path to the Zarr store.
+    codec_pipeline : CodecPipeline
+        The codec pipeline used for encoding and decoding chunks.
+    order : {'C', 'F'}
+        The order of the array data in memory.
+
+    Properties
```
[the remark this answers is not in the document]
Comment (e-marshall) [reply src/zarr/core/array.py]: Thanks @dstansby ! Making these changes

**Comment by dstansby** [comment]
Thanks a lot! I started reading the docstrings, but don't think I have the technical knowledge of the v3 codebase to properly review them, so I will leave it to someone else to do that

**Review by jhamman — APPROVED** [review]

**Pull request merged**